### PR TITLE
Feat/streamlit

### PR DESCRIPTION
**Pull request by oceans404**
# Dynamic UI demos for Testnet programs with Streamlit 🎉

Add dynamic streamlit apps for examples - https://www.loom.com/share/b61ae5f477004b8cbc7428ee9adcdb33

# Try it locally

### 1. Install dependencies

```
python3 -m venv .venv
source .venv/bin/activate
pip install -r requirements.txt
```

### 2. Create secrets file and add your nilchain private key within `.streamlit/secrets.toml`

```
cp .streamlit/secrets.toml.example .streamlit/secrets.toml
```

### 3. Run the `addition` test with starter values from `addition_test`
```
streamlit run demo.py addition_test
```

### 4. Run any other example that isn't using na.numpy or loops to define input and output parties
```
streamlit run demo.py [test-name]
```

### CODE REVIEW

**Comment by jimouris** [comment]
It might be a version mismatch but I'm getting:
```bash
TypeError: create_nillion_client() takes 2 positional arguments but 3 were given
Traceback:
File "/Users/jimouris/repos/nada-by-example/.venv/lib/python3.11/site-packages/streamlit/runtime/scriptrunner/exec_code.py", line 85, in exec_func_with_error_handling
    result = func()
             ^^^^^^
File "/Users/jimouris/repos/nada-by-example/.venv/lib/python3.11/site-packages/streamlit/runtime/scriptrunner/script_runner.py", line 576, in code_to_exec
    exec(code, module.__dict__)
File "/Users/jimouris/repos/nada-by-example/demo.py", line 196, in <module>
    main()
File "/Users/jimouris/repos/nada-by-example/demo.py", line 168, in main
    result_message = asyncio.run(store_inputs_and_run_blind_computation(input_data, program_name, output_parties, nilchain_private_key))
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/opt/homebrew/Cellar/python@3.11/3.11.9_1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/runners.py", line 190, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
File "/opt/homebrew/Cellar/python@3.11/3.11.9_1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/opt/homebrew/Cellar/python@3.11/3.11.9_1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/base_events.py", line 654, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
File "/Users/jimouris/repos/nada-by-example/nillion_client_script.py", line 25, in store_inputs_and_run_blind_computation
    client = create_nillion_client(userkey, nodekey, bootnodes)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
when I run:
```bash
streamlit run demo.py addition_test
```

File: demo.py [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+        st.caption('The Program ID is the identifier for the program you stored in the Nillion Network on the PetNet. The Program ID naming convention is your [user_id]/[program_name]')
+        st.code(result_message['program_id'], language='json')
+        
+        st.text('Nilchain Nillion Address')
+        st.caption(f"Blind computation ran on the Nillion PetNet and operations were paid for on the Nilchain Testnet. Check out the Nilchain transactions that paid for each PetNet operation (store program, store secrets, compute) on the [Nillion Testnet Explorer](https://testnet.nillion.explorers.guru/account/{result_message['nillion_address']})")
+        st.code(result_message['nillion_address'], language='json')
+        st.balloons()
+      
+if __name__ == "__main__":
+    main()
```
Comment (jimouris) [inline demo.py:196]:
```suggestion
    main()
    
```

File: .streamlit/secrets.toml.example [hunk 2]
```diff
@@ -0,0 +1 @@
+nilchain_private_key = "YOUR_PRIVATE_KEY"
```
Comment (jimouris) [inline .streamlit/secrets.toml.example:1]:
```suggestion
nilchain_private_key = "YOUR_PRIVATE_KEY"

```

File: .gitignore [hunk 3]
```diff
@@ -160,3 +160,6 @@ cython_debug/
 #  and can be added to the global gitignore or merged into this file.  For a more nuclear
 #  option (not recommended) you can uncomment the following to ignore the entire idea folder.
 #.idea/
+
+# Ignore streamlit secrets file
+.streamlit/secrets.toml
```
Comment (jimouris) [inline .gitignore:165]:
```suggestion
.streamlit/secrets.toml

```

File: demo.py [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+def main():
+    # Look for the test name to use in the demo
+    if len(sys.argv) != 2:
+        st.write("Usage: python3 demo.py <nada_test_file_name>")
+        sys.exit(1)
+
+    # Get the YAML file name from the command line argument
+    nada_test_file_name = sys.argv[1]
+
+    # Construct the YAML file path based on the provided file name
```
Comment (jimouris) [inline demo.py:110]: check here that `nada_test_file_name` exists before creating the yaml?

File: nillion_client_script.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+        elif input_type == 'SecretBoolean':
+            nada_val = nillion.SecretBoolean(value)
+        elif input_type == 'PublicBoolean':
+            nada_val = nillion.Boolean(value)
+        elif input_type == 'SecretUnsignedInteger':
+            nada_val = nillion.SecretUnsignedInteger(value)
+        elif input_type == 'PublicUnsignedInteger':
+            nada_val = nillion.UnsignedInteger(value)
+        else:
+            nada_val = nillion.Integer(value)
```
Comment (jimouris) [inline nillion_client_script.py:91]: Nit but might make it easier to maintain:
```python
# Nada types to Nillion types
types_mapping = {
    # Integers
    'SecretInteger': nillion.SecretInteger,
    'PublicInteger': nillion.Integer,
    # Unsigned Integers
    'SecretUnsignedInteger': nillion.SecretUnsignedInteger,
    'PublicUnsignedInteger': nillion.UnsignedInteger,
    # Booleans
    'SecretBoolean': nillion.SecretBoolean,
    'PublicBoolean': nillion.Boolean,
}

nada_val = types_mapping.get(input_type, nillion.Integer)(value)
```

I'm also not sure if we need to use `nillion.Integer` as the default if we don't find it in the map, maybe it'd be better to throw an error?

File: nillion_client_script.py [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+            print(compute_event)
+            print(compute_event.result)
+            blind_computation_results = compute_event.result.value
+            return {
+                'user_id': user_id,
+                'program_id': program_id,
+                'store_ids': store_ids, 
+                'output': blind_computation_results,
+                'nillion_address': payments_wallet,      
+            }
```
Comment (jimouris) [inline nillion_client_script.py:161]:
```suggestion
            }
            
```

File: requirements.txt [hunk 7]
```diff
@@ -1,2 +1,11 @@
...
+nada-numpy==0.4.0
+
+# streamlit demo dependencies
+pyyaml
+streamlit
+
+# nillion_client_script dependencies for streamlit demo
+py-nillion-client
+nillion-python-helpers
+cosmpy>=0.9.2
```
Comment (jimouris) [inline requirements.txt:11]:
```suggestion
cosmpy>=0.9.2

```

**Review by jimouris — APPROVED** [review]
For the newlines: GH shows this annoying thing if a newline is missing:
image
You can configure your editor to always add one on save.

Everything LGTM but I couldn't run it. I assume it's some version mismatch?

**Comment by oceans404** [comment]
> It might be a version mismatch but I'm getting:
> 
> ```shell
> TypeError: create_nillion_client() takes 2 positional arguments but 3 were given
> Traceback:
> File "/Users/jimouris/repos/nada-by-example/.venv/lib/python3.11/site-packages/streamlit/runtime/scriptrunner/exec_code.py", line 85, in exec_func_with_error_handling
>     result = func()
>              ^^^^^^
> File "/Users/jimouris/repos/nada-by-example/.venv/lib/python3.11/site-packages/streamlit/runtime/scriptrunner/script_runner.py", line 576, in code_to_exec
>     exec(code, module.__dict__)
> File "/Users/jimouris/repos/nada-by-example/demo.py", line 196, in <module>
>     main()
> File "/Users/jimouris/repos/nada-by-example/demo.py", line 168, in main
>     result_message = asyncio.run(store_inputs_and_run_blind_computation(input_data, program_name, output_parties, nilchain_private_key))
>                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> File "/opt/homebrew/Cellar/python@3.11/3.11.9_1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/runners.py", line 190, in run
>     return runner.run(main)
>            ^^^^^^^^^^^^^^^^
> File "/opt/homebrew/Cellar/python@3.11/3.11.9_1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/runners.py", line 118, in run
>     return self._loop.run_until_complete(task)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> File "/opt/homebrew/Cellar/python@3.11/3.11.9_1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/base_events.py", line 654, in run_until_complete
>     return future.result()
>            ^^^^^^^^^^^^^^^
> File "/Users/jimouris/repos/nada-by-example/nillion_client_script.py", line 25, in store_inputs_and_run_blind_computation
>     client = create_nillion_client(userkey, nodekey, bootnodes)
>              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> ```
> 
> when I run:
> 
> ```shell
> streamlit run demo.py addition_test
> ```

yep! I updated a dependency, nillion-python-helpers, and waiting for this PR to get merged to bump up the version in this package. I was testing locally with changes https://github.com/NillionNetwork/nillion-python-helpers/pull/6

**Pull request merged**